src/libstd/fs.rs

```diff
@@ -844,6 +844,8 @@ pub fn symlink_metadata<P: AsRef<Path>>(path: P) -> io::Result<Metadata> {
 
 /// Rename a file or directory to a new name.
 ///
+/// This will not work if the new name is on a different filesystem.
+///
```

Comment:
I believe that depends on the filesystem. The trivial example is bind mount which lets you mount a directory:
```bash
mkdir /path/to/a
mkdir /path/to/b
mount -o bind /path/to/a /path/to/b
```
These both have the same filesystem (the directory `/path/to/a`) but different mount points. See the EXDEV comment in `man 2 rename` (on linux).